### PR TITLE
[WIP][DO_NOT_MERGE] KVM implementation of tasks

### PR DESCRIPTION
**Pull request by cshari-zededa**
Signed-off-by: Hariharasubramanian C S <cshari@zededa.com>

### CODE REVIEW

File: pkg/pillar/containerd/containerd.go [hunk 1]
```diff
@@ -557,6 +557,10 @@ func CtrDelete(containerID string) error {
 	// do this just in case
 	_ = CtrStop(containerID, true)
 
+	stdinDir := path.Join("/run", "containers-stdin", containerID)
+	if err := os.RemoveAll(stdinDir); err != nil {
```
Comment (rvs) [inline pkg/pillar/containerd/containerd.go:561]: Why are you doing this?

File: pkg/pillar/containerd/containerd.go [hunk 2]
```diff
@@ -706,35 +710,43 @@ func createMountPointExecEnvFiles(containerPath string, mountpoints map[string]s
 // it R/O into the container. On top of that we expect the usual suspects of /run,
 // /persist and /config to be taken care of by the OCI config that lk produced.
 func LKTaskLaunch(name, linuxkit string, domSettings *types.DomainConfig, args []string) (int, error) {
-	config := "/containers/services" + linuxkit + "/config.json"
-	rootfs := "/containers/services" + linuxkit + "/lower"
+	config := "/containers/services/" + linuxkit + "/config.json"
+	rootfs := "/containers/services/" + linuxkit + "/rootfs"
```
Comment (rvs) [inline pkg/pillar/containerd/containerd.go:714]: If you need a R/W filesystem you need to prepare a volume for it. A better question is: why DO you need a R/W root filesystem? All the R/W bits are given to you via /run and /persist anyway -- so what's missing?

File: pkg/pillar/containerd/containerd.go [hunk 3]
```diff
@@ -706,35 +710,43 @@ func createMountPointExecEnvFiles(containerPath string, mountpoints map[string]s
...
+		log.Errorf("NewOciSpec failed with error %v", err)
 		return 0, err
 	}
 	if err = spec.Load(f); err != nil {
 		return 0, fmt.Errorf("can't load spec file from %s %v", config, err)
 	}
 
 	spec.Root.Path = rootfs
-	spec.Root.Readonly = true
 	if domSettings != nil {
```
Comment (rvs) [inline pkg/pillar/containerd/containerd.go:732]: Same question here

File: pkg/pillar/hypervisor/kvm.go [hunk 4]
```diff
@@ -300,8 +303,8 @@ func newKvm() Hypervisor {
 		return kvmContext{
 			domains:      map[string]int{},
 			devicemodel:  "pc-q35-3.1",
-			dmExec:       "qemu-system-x86_64",
-			dmArgs:       []string{"-display", "none", "-daemonize", "-S", "-no-user-config", "-nodefaults", "-no-shutdown", "-no-hpet"},
+			dmExec:       "/usr/lib/xen/bin/qemu-system-x86_64",
```
Comment (rvs) [inline pkg/pillar/hypervisor/kvm.go:306]: Why do you need to add full path here?

File: pkg/pillar/containerd/containerd.go (hunk 1 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:561]: Just cleaning it up, since the next CtrCreate expects these directories to be not present. This got exposed when LKTaskLaunch() deletes existing container (if any) before creating new container with the same domainID

File: pkg/pillar/containerd/containerd.go (hunk 2 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:714]: I had to make rootfs as R/W because of  /dev/, /sys, /proc systems, since qemu/kernel is creating /dev/pts/ and  /proc/ files during the VM launch.

File: pkg/pillar/containerd/containerd.go (hunk 3 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:732]: I had to make rootfs as R/W because of  /dev/, /sys, /proc systems, since qemu/kernel is creating /dev/pts/ and  /proc/ files during the VM launch.

File: pkg/pillar/hypervisor/kvm.go (hunk 4 above)
Comment (cshari-zededa) [reply pkg/pillar/hypervisor/kvm.go:306]: Since /usr/lib/xen/bin is not a standard $PATH, I had to mention the absolute path for finding qemu-system-x86_64 during task launch. Currently it is working because from pillar we are launching /usr/bin/qemu-system-x86_64, which actually makes the cross-container call.

File: pkg/pillar/containerd/containerd.go (hunk 1 above)
Comment (rvs) [reply pkg/pillar/containerd/containerd.go:561]: I think we need to dig into this more @cshari-zededa -- since it may expose an issue with how we're using containerd and I'd rather fix the root cause (since containerd becomes a center piece of our design). Is there an easy way to reproduce this?

File: pkg/pillar/containerd/containerd.go (hunk 2 above)
Comment (rvs) [reply pkg/pillar/containerd/containerd.go:714]: If those are the only ones -- we need to dig deeper into this as well. Those are supposed to be R/W mounted from the outside when you ask runc to run the container. Lets start with seeing what gets mounted on those endpoint when you call LKTaskLaunch() and we can proceed from there.

File: pkg/pillar/hypervisor/kvm.go (hunk 4 above)
Comment (rvs) [reply pkg/pillar/hypervisor/kvm.go:306]: Ah! Ok -- makes sense

File: pkg/pillar/containerd/containerd.go (hunk 2 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:714]: Hi @rvs, you are right, I unnecessarily marked this as R/W, whereas Readonly rootfs works just fine, as I verified it today.

File: pkg/pillar/containerd/containerd.go (hunk 1 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go]: It is not reproducible, I think I added this change unnecessarily. Removed it!

File: pkg/pillar/containerd/containerd.go (hunk 3 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:729]: reverted the Readonly change to true.

File: pkg/pillar/containerd/containerd.go (hunk 2 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:710]: Hi @rvs, I already reverted the Root.Readonly = false part.

File: pkg/pillar/containerd/containerd.go (hunk 1 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go]: Hi @rvs, 
With the [#containerd issue ](https://github.com/containerd/containerd/issues/4019) fixed, 
we no longer need this stdin workaround, so removed that piece of code, and this cleanup 
code as well. Please let me know if you think we should retain it.

File: pkg/pillar/containerd/containerd.go (hunk 2 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:697]: The only issue with read-only root partition is that, we can't use "screen" with default settings, 
since it tries to create /tmp/.screens directory, and it fails. But we can always override that by 
explicitly mentioning "SCREENDIR" env variable. So we are all good with the Read-only setting.

File: pkg/pillar/containerd/containerd.go (hunk 3 above)
Comment (cshari-zededa) [reply pkg/pillar/containerd/containerd.go:716]: The only issue with read-only root partition is that, we can't use "screen" with default settings,
since it tries to create /tmp/.screens directory, and it fails. But we can always override that by
explicitly mentioning "SCREENDIR" env variable. So we are all good with the Readonly root fs.

**Comment by cshari-zededa** [comment]
> Hey @cshari-zededa -- sorry for the delay -- lets make sure we commit it this week. A few things left to do here:
> 
> 1. rebase on the latest EVE

Done

> 2. take care of the feedback I left just now

Done, please let me know if you think I have missed anything.

> 3. lets see if we can reduce overhead to less than 800M (800M really seems excessive)
> 

Arrived at 500MB with some trail-and-error experiments with PCI passthrough settings. 

> For #3 I suggest we do the memory pinning and start passing `-overcommit mem-lock=on` and `-overcommit cpu-pm=on` to the qemu we're running.
> 

Done

> I think that will help us arrive at a more reasonable overhead figure

File: pkg/xen-tools/Dockerfile.in [hunk 5]
```diff
@@ -74,6 +74,7 @@ RUN apk add --no-cache \
     pixman=0.34.0-r5   \
     yajl=2.1.0-r0      \
     keyutils=1.5.10-r0 \
+    screen=4.6.2-r1    \
```
Comment (rvs) [inline pkg/xen-tools/Dockerfile.in:77]: Lets not pull screen in for now -- anyone debugging the system can always apk add screen (and in fact, we should probably move towards using a socket for stdio anyway -- but that shouldn't be part of this PR).

Comment (rvs) [reply pkg/xen-tools/Dockerfile.in:77]: Note that you don't have to be in the same R/O rootfs/container for using screen

Comment (cshari-zededa) [reply pkg/xen-tools/Dockerfile.in]: Hi @rvs,  excluded this change to pull in screen. I've updated the PR. Please take a look.

**Comment by rvs** [comment]
Thanks @cshari-zededa -- merged!
